magma/types/ready_valid.py

```diff
@@ -96,7 +96,7 @@ class ReadyValidProducerKind(ReadyValidKind):
     def __getitem__(cls, T: Union[Type, MagmaProtocol]):
         fields = {"valid": Out(Bit), "ready": In(Bit)}
         _maybe_add_data(cls, fields, T, Out)
-        t = type(f"{cls}[{T}]", (cls, ), fields)
+        t = type(f"{cls}[{T}]", (cls, cls.__bases__[-1][T],), fields)
```

Comment:
how does hwtypes get this to work with `Array` and `Tuple`?

Reply:
With some pretty complicated logic, here's the magma implementation: https://github.com/phanrahan/magma/blob/master/magma/tuple.py#L47-L71 which is based off the generic hwtypes implementation

Reply:
oh boy, ok. fine with this, can robust-ify it down the line if it becomes an issue.